File: src/main/webapp/js/services/subService.js
/**
 * Created by borisbondarenko on 28.12.15.
 *
 * @description
 * Service which allows a user to manage his subscribtions
 *
 * @author brzzbr
 */
(function () {

    angular.module('pinmap')
        .service('subService', ['$resource', 'notifyService',subService]);

    function subService($resource, notifyService) {

        var subResource = $resource('/api/subs/:user',
            {
                user: '@user'
            },
            {
                subscribe: {
                    method: 'POST'
                },
                unsubscribe: {
                    method: 'DELETE'
                }
            });

        return {
            getMySubs: getMySubs,
            subscribe: subscribe,
            unsubscribe: unsubscribe
        };

        /**
         * gets all the subscriptions for current user
         * @returns {*}
         */
        function getMySubs() {
            return subResource.get().$promise
                .then(function (subs) {
                    return subs;
                }, function(error){

                    console.log(error);
                    notifyService.error('We have got an error while getting your subscriptions');
                });
        }

        /**
         * allows to subscribe on user
         * @param user
         * @returns {*}
         */
        function subscribe(user) {
            return subResource.subscribe({user: user}).$promise
                .then(function (result) {
                    return result;
                }, function(error){

                    console.log(error);
                    notifyService.error('We have got an error while subscribing on a new author');
                });
        }

        /**
         * allows to unsubscribe from user
         * @param user
         * @returns {*}
         */
        function unsubscribe(user) {
            return subResource.unsubscribe({user: user}).$promise
                .then(function (result) {

                }, function(error){

                    console.log(error);
                    notifyService.error('We have got an error while unsubscribing');
                });
        }
    }
})();

File: src/main/java/com/whitesoft/pinmap/services/SubService.java
package com.whitesoft.pinmap.services;

import com.whitesoft.pinmap.domain.Sub;
import com.whitesoft.pinmap.domain.User;

import java.util.List;

/**
 * Created by borisbondarenko on 28.12.15.
 *
 * Interface for subscription service. Allows to get subscriptions,
 * subscribe or unsubscribe on users' pins.
 *
 * @author brzzbr
 */
public interface SubService {

    /**
     * Gets all subscriptions for current user
     * @return a list of user"s subscriptions
     */
    List<Sub> getSubs(User user);

    /**
     * Subscribes subscriber to author's pins
     * @param subscriber subscriber user
     * @param author author user
     */
    Sub subscribe(User subscriber, User author);

    /**
     * Unsubscribes subscriber from author's pins
     * @param subscriber subscriber user
     * @param author author user
     */
    void unsubscribe(User subscriber, User author);
}
